azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -182,12 +182,26 @@ def run(self, goal_state, ext_gs_updated):
             if not self._updater.is_retrieved_version_allowed_to_update(agent_family):
                 return
             self._updater.log_new_agent_update_message()
-            self._updater.purge_extra_agents_from_disk()
             agent = self._updater.download_and_get_new_agent(self._protocol, agent_family, goal_state)
-            if agent.is_blacklisted or not agent.is_downloaded:
-                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
-                    str(agent.version))
+
+            # We are not updating the agent if it is not downloaded properly or agent is blacklisted and already attempted 3 times.
+            # For some reason, if agent is blackslisted in previous attempts(<3), we clear the error state to make agent as good agent.
+            # If we allow to update, we increment the update attempt count.
+            if not agent.is_downloaded:
+                msg = "Agent: {0} is not downloaded properly, skipping agent update".format(str(agent.version))
                 raise AgentUpdateError(msg)
+            elif agent.get_update_attempt_count() >= 3 and agent.is_blacklisted:
```

Comment:
why the check on agent.is_blacklisted? isn't the count enough?

Reply:
well, we get this condition only when agent is blacklisted, otherwise we never multiple attempts. So I want to double check that.

Reply:
ok, if we get here only when agent.is_blacklisted is True, then it is a redundant check. why do we want to double-check?